### ISSUE TITLE
Auto-formatting C# preprocessor directives sometimes removes spaces between the tokens

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#53859]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Auto-formatting-C-preprocessor-directiv/1440195)._

---
[severity:It bothers me. A fix would be nice]
In the following code I want a space between the `!` and `DEBUG` for readability, but when I use `Ctrl+K, D` to auto-format the document it removes the space from the one around `Test4`.

```
#if ! DEBUG
public class Test { }
#endif

/// <summary></summary>
#if ! DEBUG
public class Test2 { }
#endif

namespace Namespace
{
#if ! DEBUG
    public class Test3 { }
#endif

/// <summary></summary>
#if ! DEBUG// Won't let me put a space in this one.
    public class Test4 { }
#endif
#if DEBUG
    public class Test4 { }
#endif

#if ! DEBUG
    public class Test5 { }
#endif

}

#if ! DEBUG
public class Test6 { }
#endif
```

---
#### Feedback Bot on 6/3/2021, 01:07 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by sharwell** [comment]
I don't believe we have any formatting option to allow a space between a unary operator and the operand.

**Comment by sharwell** [comment]
Note that this _only_ occurs when a `#if` line appears between a documentation comment and the element being documented.

**Comment by geniuszxy** [comment]
> 
> 
> Note that this _only_ occurs when a `#if` line appears between a documentation comment and the element being documented. Since this is a relatively new location for trivia, and none of the other locations seem to format these expressions, I'll mark this as help wanted to make this case also get ignored by the formatter.

It seems that this occurs everywhere.
Usually I'd like to indent the inner preprocess directives with spaces,
```
#if A
#    if B
...
#    endif
#endif
```
Auto formatting removes all of them now.

**Comment by SilentSin** [comment]
It doesn't occur *everywhere* for me, but I haven't figured out what the pattern is. Sometimes it happens between a documentation comment and target, sometimes it happens elsewhere when there are no comments around.

**Comment by sharwell** [comment]
@geniuszxy This issue refers to this space:

```
#if ! DEBUG
     ↑
```

Can you file a separate issue for the space between the `#` and `if`?

**Comment by CyrusNajmabadi** [comment]
Tracking with https://github.com/dotnet/roslyn/issues/31691
